AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +615,840 @@ class AzureIoTHub {
         }
 
-        // done is called everytime a message is available, params - error, data
-        function receive(done) {
-            _handlers.onEvent <- done;
+        // Closes the connection to Azure IoT Hub. Does nothing if the connection is already closed.
+        //
+        // Returns:                         Nothing.
+        function disconnect() {
+            if ((!_isDisconnected || _isConnecting) && !_isDisconnecting) {
+                _isDisconnecting = true;
+                _mqttclient.disconnect(_onDisconnect.bindenv(this));
+            }
+        }
 
-            if (done == null) {
-                _receivers.event.close();
-                // clear out handlers
-                return this;
+        // Checks if the client is connected to Azure IoT Hub.
+        //
+        // Returns:                         Boolean: true if the client is connected, false otherwise.
+        function isConnected() {
+            return _isConnected;
+        }
+
+        // TODO: update description
+        // Sends a message to Azure IoT Hub (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#sending-device-to-cloud-messages).
+        //
+        // Parameters:
+        //     msg : AzureIoTHub.Message    Message to send.
+        //     onDone : Function            Callback called when the operation is completed or an error happens.
+        //          (optional)              The callback signature:
+        //                                  onDone(error), where
+        //                                      error : Integer     0 if the operation is completed successfully, an error code otherwise.
+        //
+        // Returns:                         Nothing.
+        function sendMessage(msg, onSent = null) {
+            local tooManyRequests = _msgPendingQueue.len() >= _options.maxPendingSendRequests;
+
+            if (!_isConnected || _isDisconnecting) {
+                onSent && onSent(msg, _isConnected ? AZURE_IOT_CLIENT_ERROR_OP_NOT_ALLOWED_NOW : AZURE_IOT_CLIENT_ERROR_NOT_CONNECTED);
+                return;
             }
 
-            if( !_isOpen(_sessions.event) ) {
-                _handlers.onEvent("Cannot receive while disconnected.", null);
-            } else {
-                // TODO: test/handle if receiver already open
-                if ( _isOpen(_receivers.event) ) _receivers.event.close();
-                _openEventReceiver();
+            if (tooManyRequests) {
+                onSent && onSent(msg, AZURE_IOT_CLIENT_ERROR_OP_NOT_ALLOWED_NOW);
+                return;
             }
-            return this;
-        }
 
-        // PRIVATE FUNCTIONS
-        // ------------------------------------------------------------------------------------
-        // ------------------------------------------------------------------------------------
+            local props = "";
+            if (msg.getProperties() != null) {
```

Comment:
This is going to fail on `null`. What happens if one called `send(null);` by mistake or on purpose?

Reply:
Done